core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -717,9 +718,10 @@ public RelOptCost getCost(RelNode rel, RelMetadataQuery mq) {
   public RelSubset getSubset(RelNode rel) {
     assert rel != null : "pre: rel != null";
     if (rel instanceof RelSubset) {
-      return (RelSubset) rel;
+      return canonize((RelSubset) rel);
     } else {
-      return mapRel2Subset.get(rel);
+      RelSubset s = mapRel2Subset.get(rel);
+      return s == null ? null : canonize(s);
     }
```

Comment:
I am wondering if we can add some tests. Maybe it is diffcult.

Reply:
Yes, adding a trivial unit test is easy but not helpful, and it is quite difficult to really write it properly and e2e.